Use parameter classes instead of arguments for all physics queries

pouleyKetchoupp opened the pull request:
Same as what is already done for shape queries, applied to point and ray queries. Easier to document and more flexible to add more parameters.

Also expose `intersect_point` method to script in 3D.
(closes https://github.com/godotengine/godot-proposals/issues/3262 for 4.0)

Remove `intersect_point_on_canvas` in 2D, replaced with a parameter.

pouleyKetchoupp committed:
Use parameter classes instead of arguments for all physics queries

Same as what is already done for shape queries, applied to point and ray
queries. Easier to document and more flexible to add more parameters.

Also expose intersect_point method to script in 3D.
Remove intersect_point_on_canvas in 2D, replaced with a parameter.

fabriceci commented on servers/physics_server_2d.cpp:
```diff
@@ -134,95 +134,132 @@ PhysicsDirectBodyState2D::PhysicsDirectBodyState2D() {}
+	ADD_PROPERTY(PropertyInfo(Variant::VECTOR2, "position"), "set_position", "get_position");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "canvas_instance_id", PROPERTY_HINT_OBJECT_ID), "set_canvas_instance_id", "get_canvas_instance_id");
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "collision_mask", PROPERTY_HINT_LAYERS_2D_PHYSICS), "set_collision_mask", "get_collision_mask");
+	ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "exclude", PROPERTY_HINT_NONE, itos(Variant::RID) + ":"), "set_exclude", "get_exclude");
```
Just for info, could you explain to me what does this `itos(Variant::RID) + ":"` in the PropertyInfo?

pouleyKetchoupp replied:
Oh, thanks! I had just copied these lines from PhysicsShapeQueryParameters and I have no idea what it's supposed to do :D

It looks like it's the only case in the engine where an Array of a specific type is done this way, so I'm going to change it to the proper syntax, which seems to be `PROPERTY_HINT_ARRAY_TYPE, "RID"`.

pouleyKetchoupp commented:
Updated from @fabriceci's comment and made a couple more fixes around `intersect_shape`:
In 2D, I've made a little fix that allows `motion` to work properly and removed the note in doc that says it doesn't.
In 3D, this is not supported (yet) so I've added a similar note in the doc.

pouleyKetchoupp commented:
And I've pushed one last fix to add the missing `motion` property in `PhysicsShapeQueryParameters3D`, since now it's working the same way as in 2D (shape motion in parameters rather than argument).

Should be good to go now :)

akien-mga commented:
Thanks!
